[clang][ARM] Fix warning for using VFP from interrupts.

chrisnc opened the pull request:
[clang][ARM] Fix warning for using VFP from interrupts.

This warning has three issues:
 - The interrupt attribute causes the function to return using an exception
   return instruction. This warning allows calls from one function with
   the interrupt attribute to another, and the diagnostic text suggests
   that not having the attribute on the callee is a problem. Actually
   making such a call will lead to a double exception return, which is
   unpredictable according to the ARM architecture manual section
   B9.1.1, "Restrictions on exception return instructions". Even on
   machines where an exception return from user/system mode is
   tolerated, if the callee's interrupt type is anything other than a
   supervisor call or secure monitor call, it will also return to a
   different address than a normal function would. For example,
   returning from an "IRQ" handler will return to lr - 4, which will
   generally result in calling the same function again.
 - The interrupt attribute currently does not cause caller-saved VFP
   registers to be saved and restored if they are used, so putting
   __attribute__((interrupt)) on a called function doesn't prevent it
   from clobbering VFP state.
 - It is part of the -Wextra diagnostic group and can't be individually
   disabled when using -Wextra, which also means the diagnostic text of
   this specific warning appears in the documentation of -Wextra.

This change addresses all three issues by instead generating a warning
for any interrupt handler where the vfp feature is enabled. The warning is
also given its own diagnostic group.

Closes #34876.

[clang][ARM] Emit an error when an interrupt handler is called.

Closes #95359.

chrisnc committed:
[clang][ARM] Fix warning for using VFP from interrupts.

This warning has three issues:
 - The interrupt attribute causes the function to return using an exception
   return instruction. This warning allows calls from one function with
   the interrupt attribute to another, and the diagnostic text suggests
   that not having the attribute on the callee is a problem. Actually
   making such a call will lead to a double exception return, which is
   unpredictable according to the ARM architecture manual section
   B9.1.1, "Restrictions on exception return instructions". Even on
   machines where an exception return from user/system mode is
   tolerated, if the callee's interrupt type is anything other than a
   supervisor call or secure monitor call, it will also return to a
   different address than a normal function would. For example,
   returning from an "IRQ" handler will return to lr - 4, which will
   generally result in calling the same function again.
 - The interrupt attribute currently does not cause caller-saved VFP
   registers to be saved and restored if they are used, so putting
   __attribute__((interrupt)) on a called function doesn't prevent it
   from clobbering VFP state.
 - It is part of the -Wextra diagnostic group and can't be individually
   disabled when using -Wextra, which also means the diagnostic text of
   this specific warning appears in the documentation of -Wextra.

This change addresses all three issues by instead generating a warning
for any interrupt handler where the vfp feature is enabled. The warning is
also given its own diagnostic group.

Closes #34876.

chrisnc commented:
Here is an example of the existing warning in action. https://godbolt.org/z/9e84EfeYP
It warns for calling a normal function, but the same function with an interrupt attribute does not warn. The `subs pc, lr, 4` in `bar_irq` goes back to `bl bar_irq`, creating an infinite loop, where the second `subs pc, lr, 4` from `bar_irq` is also unpredictable, because the processor mode will most likely be system/user.

chrisnc commented:
I'd like to also address #47815, but I wasn't able to find the right path to answer "is this M-profile" from `TargetInfo`. Would appreciate any suggestions on that. I think it may require `dynamic_cast`-ing the `TargetInfo` to an `ARMTargetInfo`, and then change one of the methods in there to be public...

chrisnc commented:
ping @ostannard @smithp35

chrisnc commented:
Rebased and fixed conflict in release notes.

chrisnc commented:
ping (rebased and fixed another release notes conflict)

chrisnc commented:
Ping

chrisnc commented:
Ping @DavidSpickett @jthackray

DavidSpickett commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -544,6 +548,14 @@ Improvements to Clang's diagnostics
 - Clang no longer emits a "declared here" note for a builtin function that has no declaration in source.
   Fixes #GH93369.
 
+- On ARM, Clang no longer suggests adding ``__attribute__((interrupt))`` to
```
I would phrase this "For the ARM target", since Clang could be running on any host architecture.

DavidSpickett commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -384,6 +384,10 @@ Modified Compiler Flags
   evaluating to ``true`` and an empty body such as ``while(1);``)
   are considered infinite, even when the ``-ffinite-loop`` flag is set.
 
+- Removed "arm interrupt calling convention" warning that was included in
```
I would rephrase this a bit:
```
The "arm interrupt calling convention" warning previously included in ``-Wextra`` has been removed and made into a unique warning enabled using ``-Warm-interrrupt-vfp-clobber``.
```

chrisnc replied:
"Made into" implies that it's the more or less the same warning, but I'm trying to convey that the existing warning was removed because it's unsound, and the name is part of its confusedness. I will revisit this.

DavidSpickett commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -336,9 +336,10 @@ def warn_anyx86_excessive_regsave : Warning<
-   "call to function without interrupt attribute could clobber interruptee's VFP registers">,
-   InGroup<Extra>;
+def warn_arm_interrupt_vfp_clobber : Warning<
+   "calling a VFP-enabled function from an interrupt could clobber the "
```
Unless "VFP-enabled" is some Arm defined, specific spelling, I would write it as "VFP enabled" without the "-".

chrisnc replied:
It's a phrasal adjective which would normally be hyphenated, but I will consider how to rephrase so it's not needed.

DavidSpickett reviewed (commented):
I can't fully approve this, I leave that to other members of @llvm/pr-subscribers-arm .

2 overall comments:
* Calling a function marked interrupt from a function marked interrupt is still UB isn't it? Should there be another warning to cover that scenario as well?
* I see that gcc has a warning in `-Wattributes` for this, you should raise a bug there equivalent to the one llvm has, perhaps there even is one already.

DavidSpickett commented:
> I'd like to also address https://github.com/llvm/llvm-project/issues/47815, but I wasn't able to find the right path to answer "is this M-profile" from TargetInfo.

Any ARM folks know if there are existing M profile only warnings to learn from?

chrisnc commented:
Thank you for the review!

> * Calling a function marked interrupt from a function marked interrupt is still UB isn't it? Should there be another warning to cover that scenario as well?

Correct. I can create a separate issue to describe such a warning.

> * I see that gcc has a warning in `-Wattributes` for this, you should raise a bug there equivalent to the one llvm has, perhaps there even is one already.

It does, but `gcc`'s `-Wattributes` doesn't suffer from any of the issues being addressed here: it's already a separate flag, it cannot be silenced by putting `__attribute__((interrupt))` on the callee, and its warning text describes compiling with `-mgeneral-regs-only` to avoid the issue, though this only works when passed on the command line, not with `__attribute__((target("general-regs-only")))`.

It seems that LLVM does not have this target feature for 32-bit Arm, so I opted for `soft-float` as the escape hatch, which should achieve the same result, and check for it in function attributes. It is not checked transitively though, so this PR will not catch cases where an interrupt calls a soft-float function that then calls a function that uses VFP. The latter call is otherwise allowed though. IMO this is still a lot better than not having the escape hatch, but I'm open to suggestions.

chrisnc commented:
I've rebased, updated the warning text and release notes, and created https://github.com/llvm/llvm-project/issues/95359 for the future improvement to warn about calling interrupt handlers.

chrisnc commented:
While doing some more research and comparing `gcc` and `clang`, I noticed that `clang` doesn't even save the volatile fp registers in an interrupt handler when it uses them directly, so there's no point in just checking for function calls. `gcc` does except for `fpscr`, which it will still clobber if floating-point conditionals are computed. Maybe the behavior was different when this warning was introduced, but as of now, using fp at all in an interrupt handler will clobber vfp state, so I think the right path here is to just implement `-Wattributes` in essentially the same way `gcc` does...

chrisnc committed:
[clang][ARM] Emit an error when an interrupt handler is called.

Closes #95359.

chrisnc commented:
Updated to just warn if the attribute is used while vfp is enabled, and added an error for calling an interrupt handler, similar to x86.

evodius96 commented:
Somewhat related to this -- we (TI) are presently in the process of upstreaming our mod to save/restore VFP registers with the interrupt_save_fp attribute. See #89654.  I think the patch needs to be updated in light of upstream changes to the Arm frame lowering code.

chrisnc commented:
Rebased to fix release not conflict. @DavidSpickett @llvm/pr-subscribers-arm

DavidSpickett commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -336,9 +336,12 @@ def warn_anyx86_excessive_regsave : Warning<
+  "interruptee's vfp state">,
+  InGroup<DiagGroup<"arm-interrupt-vfp-clobber">>;
+def err_arm_interrupt_called : Error<
+  "interrupt service routine cannot be called directly">;
```
I don't see a test case for this error.

chrisnc replied:
Added.

DavidSpickett commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -448,6 +448,11 @@ Modified Compiler Flags
+- Removed "arm interrupt calling convention" warning that was included in
+  ``-Wextra`` without its own flag.
+
+- Added ``-Warm-interrupt-vfp-clobber``, with its own warning group.
```
We should be very clear whether this warning is serving the same function as the one that was removed, or not.

If it is, perhaps it is better to merge the two points so it is clear one is replaced by the other. If not, perhaps you can include a sentence for each like "that would be emitted when..." - "that will be emitted when..." that shows the purpose of each one.

(I know this is in the help, but we can save interested parties a round trip here)

chrisnc replied:
Moved these around, and split up the content that was previously in the `Improvements to Clang's diagnostics` section.

chrisnc commented:
Clarified the release notes changes and added a test case for the new error.

chrisnc commented:
Fixed another release notes conflict.

Ping @DavidSpickett @ostannard @jthackray.

DavidSpickett commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -1329,6 +1329,11 @@ void SemaARM::handleInterruptAttr(Decl *D, const ParsedAttr &AL) {
+  const TargetInfo &TI = getASTContext().getTargetInfo();
+  if (TI.hasFeature("vfp")) {
+    Diag(D->getLocation(), diag::warn_arm_interrupt_vfp_clobber);
+  }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

chrisnc replied:
Huh, I guess the rule has exceptions so the clang-format checker doesn't flag it? Will fix.

DavidSpickett replied:
I think clang-format will very rarely add or remove tokens, just shuffle them around. So unfortunately this "rule" is just a when we see it kinda thing.

chrisnc replied:
Oh, right. Well, fixed now.

DavidSpickett commented:
Everything LGTM but I'd like someone who works on Arm's downstream compiler to approve. @ostannard perhaps? 

(thanks for your patience @chrisnc)

DavidSpickett commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -492,6 +495,13 @@ Modified Compiler Flags
+  handlers to prevent clobbering VFP registers. Following this suggestion leads
+  to unpredictable behavior by causing multiple exception returns from one
+  exception. Fixes #GH34876.
+
```
Note to other reviewers: This is in the modified flags section because -wextra itself is being modified. A part of it was removed but that part was never a separate flag so this doesn't need to go in the "removed compiler flags" section.

ostannard reviewed (approved):
LGTM

chrisnc commented:
Force-pushed to fix the single-line `if`; did not rebase so builds should still be passing. Ready to merge (need someone else to).

chrisnc commented:
Thanks!

ahmedbougacha commented:
> > * Calling a function marked interrupt from a function marked interrupt is still UB isn't it? Should there be another warning to cover that scenario as well?
> 
> Correct. I can create a separate issue to describe such a warning. Calling an interrupt handler from a normal function is also UB for similar reasons; when it returns it will attempt an exception return for an exception that hasn't occurred, so either it's immediately undefined because the caller was already in system/user, or the caller was in supervisor mode from reset, and an exception return will update CPSR to whatever garbage was in SPSR_svc, and also it will return to the calling instruction, rather than one past it.

Catching up with this: as far as I can tell this is true for A/R profiles, but M profile doesn't have this restriction, because the exception return sequence isn't special.  And the plain return we codegen only behaves as an exception return when the LR bits are setup in the special way.  So it can return just fine if LR is a regular return address.  I read that as saying there's no restriction to calling `interrupt` functions when targeting M profile.  Does that sound correct to you all?  Is there another reason to restrict calls even for M profile?  (beyond "why would anyone ever do that" ;)

I see you mentioned the related #47815 :
> I'd like to also address https://github.com/llvm/llvm-project/issues/47815, but I wasn't able to find the right path to answer "is this M-profile" from TargetInfo. Would appreciate any suggestions on that. I think it may require dynamic_cast-ing the TargetInfo to an ARMTargetInfo, and then change one of the methods in there to be public...

I think the somewhat-clean way to do that would be to add a bit in `TargetInfo` (see e.g., `ARMCDECoprocMask`), initialize it in `ARMTargetInfo` based on `ArchProfile == M`, and checking it in Sema via `Context.getTargetInfo()`

chrisnc commented:
@ahmedbougacha yes, I think it's correct that calling an `__attribute__((interrupt))` function is safe to do on M-profile because there is no difference in the return sequence; all it does is not assume that the stack is 8-byte aligned initially. Even then, this is only needed when not using the STKALIGN bit in CCR, which is on by default on most (all?) Cortex-M processors. This feature, along with the automatic volatile register saving, mean that `__attribute__((interrupt))` has almost no legitimate uses on M-profile.
